test_plans/any_device_has_all_devices_have.asciidoc

```diff
@@ -0,0 +1,42 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for traits `any_device_has` and `all_devices_have`
+
+This is a test plan for traits that the application can use to query aspects at compilation time
+that are described in SYCL 2020 https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:device-aspects[4.6.4.3. Device aspects].
+
+== Testing scope
+
+=== Aspect coverage
+
+Tests described below are performed for following aspects:
+
+* `sycl::aspect::cpu`
+* `sycl::aspect::gpu`
+* `sycl::aspect::accelerator`
+* `sycl::aspect::custom`
+* `sycl::aspect::emulated`
+* `sycl::aspect::host_debuggable`
+* `sycl::aspect::fp16`
+* `sycl::aspect::fp64`
+* `sycl::aspect::atomic64`
+* `sycl::aspect::image`
+* `sycl::aspect::online_compiler`
+* `sycl::aspect::online_linker`
+* `sycl::aspect::queue_profiling`
+* `sycl::aspect::usm_device_allocations`
+* `sycl::aspect::usm_host_allocations`
+* `sycl::aspect::usm_atomic_host_allocations`
+* `sycl::aspect::usm_shared_allocations`
+* `sycl::aspect::usm_atomic_shared_allocations`
+* `sycl::aspect::usm_system_allocations`
+
+== Tests
+
+* Check that `any_device_has<aspect>` inherits from `std::true_type` if there is any device available that supports tested aspect,
+and inherits from `std::false_type` otherwise
```

Comment:
I think this test is not valid.  What exactly do you mean by:

> if there is any device available that supports tested aspect

I assume that you plan to run the program and test whether any of the devices that are available on the test system have this aspect?

This is not a valid test.  The spec says:

> The trait `any_device_has<aspect>` inherits from `std::true_type` if the **compilation environment** supports any device which has the specified aspect

(Emphasis mine.)

It is possible that the compilation environment could support some devices that are not present on the system where you run the test program.

Therefore, I think you need to weaken the test to this:

* Check each device to see if it has aspect _A_.
* If some device has aspect _A_ and `any_device_has_v<A>` is `false`, then the test fails.
* If some device does not have aspect _A_ and `all_devices_have_v<A>` is `true`, then the test fails.

Reply:
Thank you! Fixed test plan.